### PR TITLE
feat: add support for ttl after finished controller for cleaning up finished jobsets

### PR DESCRIPTION
**Pull request by dejanzele**
**CHANGELOG**
* Added `TTLSecondsAfterFinished *int32` in `JobSetSpec` with validation for minimal value 0 and default value 0
* Added TTLAfterFinished controller which watches for finished JobSets with configured TTL, similarly how it is implemented in Kubernetes for Job controller
* Added unit and integration tests
* Edited Dockerfile to include client-go folder in build container

This PR closes https://github.com/kubernetes-sigs/jobset/issues/279

### CODE REVIEW

**Comment by dejanzele** [comment]
@kannon92 @danielvegamyhre as I have issues running JobSet API locally on my M1, can you please help me verify this works?

File: test/integration/controller/ttlafterfinished_controller_test.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+		gomega.Expect(k8sClient.Get(ctx, client.ObjectKeyFromObject(js), &fresh)).To(gomega.Succeed())
+		gomega.Expect(fresh.DeletionTimestamp).To(gomega.BeNil())
+
+		ginkgo.By("checking that ttl after finished controller deletes jobset after configured seconds pass")
+		gomega.Eventually(func() bool {
+			if err := k8sClient.Get(ctx, client.ObjectKeyFromObject(js), &fresh); err != nil {
+				return false
+			}
+			return !fresh.DeletionTimestamp.IsZero()
+		}, 15*time.Second, 500*time.Millisecond).Should(gomega.BeTrue())
```
Comment (kannon92) [inline test/integration/controller/ttlafterfinished_controller_test.go:82]: Could you also add a test case to verify jobs and pods are also deleted?

Comment (dejanzele) [reply test/integration/controller/ttlafterfinished_controller_test.go:82]: Shouldn't that be tested in the JobSet controller, as it is the responsibility of that controller? TTL After Finished controlled shouldn't know or care about cleanup logic besides marking JobSet for deletion.

**Review by danielvegamyhre — COMMENTED** [review]
This looks great! Thanks for working on this. Glad you added metrics for the TTL deletion duration as well, that is a nice addition. 

I added some comments, once those are addressed I'll take another look since this is a pretty large PR. 

Also, have you manually tested theses changes, or just used unit tests and integration tests so far?

File: main.go [hunk 2]
```diff
@@ -157,6 +163,22 @@ func setupControllers(mgr ctrl.Manager, certsReady chan struct{}) {
 		os.Exit(1)
 	}
 
+	clientset := versioned.NewForConfigOrDie(mgr.GetConfig())
+	sharedInformers := externalversions.NewSharedInformerFactory(clientset, 30*time.Minute)
```
[the remark this answers is not in the document]
Comment (ahg-g) [reply main.go:167]: We don't need to use informerfactory, kube-builder sets it up already under the client we get from the manager (`mgr.GetClient()`)

**Review by ahg-g — COMMENTED** [review]
Thanks for the PR, but kube-builder offers most the mechanics implemented in ttlafterfinished_controller.go file. I think the queue management stuff are not necessary. 

Why not use the JobSet reconciler itself? Simply queue completed jobsets for the time remaining to delete it (using `return ctrl.Result{RequeueAfter: timeToDelete} `)

**Comment by dejanzele** [comment]
@ahg-g I based it on the TTL After Finished controller for Jobs in Kubernetes.
I agree it makes sense to reuse more logic from kubebuilder. I'll refactor the PR in this way.

**Comment by ahg-g** [comment]
Hi @dejanzele, are you still working on this?

**Comment by ahg-g** [comment]
closing this due to inactivity

**Comment by ahg-g** [comment]
@dejanzele sounds good, looking forward to it.

**Pull request closed without merging**